File: frontend/src/lib/checkers.ts

import { labelNiceClass } from "@/lib/niceClasses";

export type AvailabilityStatus = "available" | "taken" | "unknown" | "similar";

export type TrademarkMatch = {
  id: string;
  words?: string;
  status?: string;
  classes?: string[];
  classLabels?: string[];
};

export type CheckResult = {
  label: string;
  status: AvailabilityStatus;
  summary?: string;
  whyThisMatters?: string;
  details?: string;
  exactMatches?: TrademarkMatch[];
  similarMatches?: TrademarkMatch[];
};

export type AggregatedResults = {
  businessName: CheckResult;
  trademark: CheckResult;
  domains: CheckResult[];
  socials: CheckResult[];
};

function normalizeQuery(name: string) {
  return name.trim().replace(/\s+/g, " ");
}

// ---- OAuth token caching (valid ~1 hour) ----
let cachedToken: { token: string; expiresAt: number } | null = null;

async function getIpauAccessToken(): Promise<string> {
  const env = process.env.IPAU_ENV || "test";

  const tokenUrl =
    env === "production"
      ? process.env.IPAU_TOKEN_URL_PROD
      : process.env.IPAU_TOKEN_URL_TEST;

  const clientId = process.env.IPAU_CLIENT_ID;
  const clientSecret = process.env.IPAU_CLIENT_SECRET;

  if (!tokenUrl) throw new Error("Missing IPAU_TOKEN_URL_TEST/PROD in .env.local");
  if (!clientId) throw new Error("Missing IPAU_CLIENT_ID in .env.local");
  if (!clientSecret) throw new Error("Missing IPAU_CLIENT_SECRET in .env.local");

  // If we have a valid cached token, reuse it
  if (cachedToken && Date.now() < cachedToken.expiresAt) {
    return cachedToken.token;
  }

  const form = new URLSearchParams();
  form.set("grant_type", "client_credentials");
  form.set("client_id", clientId);
  form.set("client_secret", clientSecret);

  const res = await fetch(tokenUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      Accept: "application/json",
    },
    body: form.toString(),
    cache: "no-store",
  });

  const text = await res.text();
  if (!res.ok) {
    throw new Error(`Token request failed ${res.status}: ${text.slice(0, 250)}`);
  }

  const data = JSON.parse(text);
  const accessToken = data?.access_token;
  const expiresIn = Number(data?.expires_in ?? 3600);

  if (!accessToken) {
    throw new Error(`Token response missing access_token: ${text.slice(0, 250)}`);
  }

  // Cache it with a small safety buffer (60s)
  cachedToken = {
    token: accessToken,
    expiresAt: Date.now() + (expiresIn - 60) * 1000,
  };

  return accessToken;
}
function extractNiceClasses(tm: any): string[] {
    // Try common shapes across APIs
    const candidates: any[] = [];
  
    // direct arrays
    if (Array.isArray(tm?.classes)) candidates.push(...tm.classes);
    if (Array.isArray(tm?.niceClasses)) candidates.push(...tm.niceClasses);
    if (Array.isArray(tm?.classifications)) candidates.push(...tm.classifications);
  
    // nested common shapes
    if (Array.isArray(tm?.goodsAndServices?.classes)) candidates.push(...tm.goodsAndServices.classes);
    if (Array.isArray(tm?.goodsAndServices?.niceClasses)) candidates.push(...tm.goodsAndServices.niceClasses);
  
    // sometimes it’s goodsAndServices items, each with a class number
    if (Array.isArray(tm?.goodsAndServices)) candidates.push(...tm.goodsAndServices);
  
    // sometimes “classifications” is object with list inside
    if (Array.isArray(tm?.classificationList)) candidates.push(...tm.classificationList);
  
    // Convert to class numbers
    const numbers = candidates
      .map((c) => (c?.number ?? c?.classNumber ?? c?.niceClass ?? c?.class ?? c?.classId ?? c))
      .map((v) => (v == null ? "" : String(v).trim()))
      .filter(Boolean);
  
    // Deduplicate + sort numerically when possible
    const unique = Array.from(new Set(numbers));
    unique.sort((a, b) => Number(a) - Number(b));
  
    return unique;
  }  

async function fetchTrademarkDetails(id: string, token: string, baseUrl: string) {
    const url = `${baseUrl.replace(/\/$/, "")}/trade-mark/${encodeURIComponent(id)}`;
  
    const res = await fetch(url, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${token}`,
        Accept: "application/json",
      },
      cache: "no-store",
    });
  
    const text = await res.text();
    if (!res.ok) {
      return { id, error: `Detail fetch failed ${res.status}: ${text.slice(0, 120)}` };
    }
  
    try {
      return { id, data: JSON.parse(text) };
    } catch {
      return { id, error: `Detail fetch returned non-JSON: ${text.slice(0, 120)}` };
    }
  }
  
// ---- IP Australia Trademark quick search ----
export async function checkIpAustraliaTrademark(name: string): Promise<CheckResult> {
    const q = normalizeQuery(name);
  
    const env = process.env.IPAU_ENV || "test";
    const baseUrl =
      env === "production"
        ? process.env.IPAU_TM_BASE_URL_PROD
        : process.env.IPAU_TM_BASE_URL_TEST;
  
    if (!baseUrl) {
      return {
        label: "Trademark (IP Australia)",
        status: "unknown",
        summary: "We couldn’t run the trademark check right now.",
        whyThisMatters:
          "Trademark results help you avoid picking a name that could conflict with an existing brand.",
        details: "Missing IPAU_TM_BASE_URL_TEST/PROD in .env.local",
        exactMatches: [],
        similarMatches: [],
      };
    }
  
    const url = `${baseUrl.replace(/\/$/, "")}/search/quick`;
  
    // ✅ Matches ApiQuickSearchRequest from the docs
    const payload = {
      query: q,
      sort: {
        field: "NUMBER" as const,
        direction: "ASCENDING" as const,
      },
      filters: {
        quickSearchType: ["WORD"] as const,
        status: ["REGISTERED"] as const,
      },
      // Optional:
      // changedSinceDate: "2019-01-15",
    };
  
    try {
      const token = await getIpauAccessToken();
  
      // 1) Quick search
      const res = await fetch(url, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${token}`,
          "Content-Type": "application/json",
          Accept: "application/json",
        },
        body: JSON.stringify(payload),
        cache: "no-store",
      });
  
      const text = await res.text();
      if (!res.ok) {
        return {
          label: "Trademark (IP Australia)",
          status: "unknown",
          summary: "We couldn’t complete the trademark check right now.",
          whyThisMatters:
            "Trademark results help you avoid choosing a name that could conflict with an existing brand.",
          details: `IP Australia error ${res.status}: ${text.slice(0, 250)}`,
          exactMatches: [],
          similarMatches: [],
        };
      }
  
      const data = JSON.parse(text);
  
      // 2) Parse quick search result
      const count = Number(data?.count ?? 0);
      const trademarkIds: string[] = Array.isArray(data?.trademarkIds) ? data.trademarkIds : [];
  
      if (count === 0 || trademarkIds.length === 0) {
        return {
          label: "Trademark (IP Australia)",
          status: "available",
          summary: "No registered word trademarks were found in Australia.",
          whyThisMatters:
            "This lowers risk, but it’s not a guarantee. Similar marks, pending applications, or other rights may still exist.",
          details: "0 registered results returned by IP Australia quick search.",
          exactMatches: [],
          similarMatches: [],
        };
      }
  
      // 3) Fetch details for top IDs (so we can show names + classes)
      const topIds = trademarkIds.slice(0, 5);
      const detailResults = await Promise.all(
        topIds.map((id) => fetchTrademarkDetails(id, token, baseUrl))
      );
  
      // 4) Normalize detailed records -> structured matches
      const qLower = q.toLowerCase();
  
      const parsed = detailResults.map((d) => {
        if ("error" in d) {
          return {
            id: d.id,
            words: "",
            status: "",
            classes: [] as string[],
            classLabels: [] as string[],
            error: true,
          };
        }
  
        const tm = (d as any).data;
        console.log("goodsAndServices sample:", JSON.stringify(tm?.goodsAndServices, null, 2).slice(0, 2000));
  
        // words/name fields vary
        const words =
          (tm?.words ??
            tm?.tradeMarkWords ??
            tm?.markText ??
            tm?.text ??
            tm?.name ??
            tm?.tradeMarkName ??
            "") as string;
  
        // status fields vary
        const status =
          (tm?.statusGroup ??
            tm?.statusCode ??
            tm.statusDetail ??
            tm?.status ??
            tm?.tradeMarkStatus ??
            tm?.state ??
            tm?.ipRightStatus ??
            "") as string;
  
        // classes/categories can be nested; try common shapes
        
        // ✅ IP Australia detail shape: goodsAndServices is an array of { class: "35", descriptionText: [...] }
        const classesSet = new Set<string>();

        if (Array.isArray(tm?.goodsAndServices)) {
            for (const gs of tm.goodsAndServices) {
                const cls = gs?.class;
                if (cls) classesSet.add(String(cls));
            }
        }

        const classes = Array.from(classesSet).sort((a, b) => Number(a) - Number(b));
        const classLabels = classes.map(labelNiceClass);

        // Optional debug (recommended while testing)
        console.log("TM classes:", classes);
        console.log("TM goodsAndServices sample:", JSON.stringify(tm?.goodsAndServices, null, 2).slice(0, 600));

        return {
            id: d.id,
            words: (words ?? "").toString(),
            status: (status ?? "").toString(),
            classes,
            classLabels,
            error: false,
        };

      });
  
      // Split into exact vs similar (ignore “error” rows)
      const exactMatches = parsed
        .filter((p) => !p.error && p.words.toLowerCase() === qLower)
        .map((p) => ({
          id: p.id,
          words: p.words,
          status: p.status,
          classes: p.classes,
          classLabels: p.classLabels,
        }));
  
      const similarMatches = parsed
        .filter((p) => !p.error && p.words.toLowerCase() !== qLower)
        .map((p) => ({
          id: p.id,
          words: p.words,
          status: p.status,
          classes: p.classes,
          classLabels: p.classLabels,
        }));
  
      // If we couldn’t parse any details, fall back to IDs only
      const anyParsed = exactMatches.length + similarMatches.length > 0;
  
      if (!anyParsed) {
        return {
          label: "Trademark (IP Australia)",
          status: "similar",
          summary: "Registered trademarks exist with this name or something similar.",
          whyThisMatters:
            "Even if names aren’t identical, similar marks in related categories can increase legal and branding risk.",
          details: `Found ${count} registered match(es). IDs: ${trademarkIds.slice(0, 5).join(", ")}`,
          exactMatches: [],
          similarMatches: trademarkIds.slice(0, 5).map((id) => ({ id })),
        };
      }
  
      const isTaken = exactMatches.length > 0;
  
      return {
        label: "Trademark (IP Australia)",
        status: isTaken ? "taken" : "similar",
        summary: isTaken
          ? "An identical registered trademark exists in Australia."
          : "Similar registered trademarks exist in Australia.",
        whyThisMatters: isTaken
          ? "Using the same name can create a high risk of trademark conflict, especially in related industries."
          : "Even if names aren’t identical, similar marks in related categories can still cause legal and branding risk.",
        details: `Found ${count} registered match(es). Showing the top ${topIds.length}.`,
        exactMatches,
        similarMatches,
      };
    } catch (err: any) {
      return {
        label: "Trademark (IP Australia)",
        status: "unknown",
        summary: "We couldn’t run the trademark check right now.",
        whyThisMatters:
          "Trademark results help you avoid choosing a name that could conflict with an existing brand.",
        details: `Request failed: ${err?.message || "unknown error"}`,
        exactMatches: [],
        similarMatches: [],
      };
    }
  }
  
// ---- MVP aggregated results ----
export async function getDemoResults(name: string): Promise<AggregatedResults> {
  const normalized = name.toLowerCase().trim();
  const compact = normalized.replace(/\s+/g, "");

  // Demo logic (keep for now)
  const isProbablyTaken = normalized.includes("koala") || normalized.includes("australia");
  const hasSimilar = normalized.includes("brew") || normalized.includes("coffee");

  // Real trademark check
  const trademark = await checkIpAustraliaTrademark(name);

  return {
    businessName: {
      label: "ASIC business name (AU)",
      status: isProbablyTaken ? "taken" : "available",
      details: isProbablyTaken
        ? "A similar or identical business name appears to exist (demo)."
        : "No exact match found (demo).",
    },
    trademark,
    domains: [
      {
        label: `${compact}.com`,
        status: isProbablyTaken ? "taken" : "available",
        details: isProbablyTaken ? "Likely registered already (demo)." : "Appears free (demo).",
      },
      {
        label: `${compact}.com.au`,
        status: hasSimilar ? "similar" : "available",
        details: hasSimilar ? "Similar domains may exist (demo)." : "Appears free (demo).",
      },
    ],
    socials: [
      {
        label: `@${compact} (Instagram)`,
        status: isProbablyTaken ? "taken" : "available",
        details: isProbablyTaken ? "Handle looks popular (demo)." : "Appears free (demo).",
      },
      {
        label: `@${compact} (TikTok)`,
        status: "available",
        details: "Appears free (demo).",
      },
    ],
  };
}

File: backend/checkers.py
"""
Brand checking logic ported from TypeScript.
"""
import os
import re
import time
import asyncio
from typing import Dict, List, Optional, Literal
import httpx
from nice_classes import label_nice_class

AvailabilityStatus = Literal["available", "taken", "unknown", "similar"]


class TrademarkMatch:
    def __init__(
        self,
        id: str,
        words: Optional[str] = None,
        status: Optional[str] = None,
        classes: Optional[List[str]] = None,
        class_labels: Optional[List[str]] = None,
    ):
        self.id = id
        self.words = words
        self.status = status
        self.classes = classes or []
        self.class_labels = class_labels or []

    def to_dict(self):
        return {
            "id": self.id,
            "words": self.words,
            "status": self.status,
            "classes": self.classes,
            "classLabels": self.class_labels,
        }


class CheckResult:
    def __init__(
        self,
        label: str,
        status: AvailabilityStatus,
        summary: Optional[str] = None,
        why_this_matters: Optional[str] = None,
        details: Optional[str] = None,
        exact_matches: Optional[List[TrademarkMatch]] = None,
        similar_matches: Optional[List[TrademarkMatch]] = None,
    ):
        self.label = label
        self.status = status
        self.summary = summary
        self.why_this_matters = why_this_matters
        self.details = details
        self.exact_matches = exact_matches or []
        self.similar_matches = similar_matches or []

    def to_dict(self):
        return {
            "label": self.label,
            "status": self.status,
            "summary": self.summary,
            "whyThisMatters": self.why_this_matters,
            "details": self.details,
            "exactMatches": [m.to_dict() for m in self.exact_matches],
            "similarMatches": [m.to_dict() for m in self.similar_matches],
        }


class AggregatedResults:
    def __init__(
        self,
        business_name: CheckResult,
        trademark: CheckResult,
        domains: List[CheckResult],
        socials: List[CheckResult],
    ):
        self.business_name = business_name
        self.trademark = trademark
        self.domains = domains
        self.socials = socials

    def to_dict(self):
        return {
            "businessName": self.business_name.to_dict(),
            "trademark": self.trademark.to_dict(),
            "domains": [d.to_dict() for d in self.domains],
            "socials": [s.to_dict() for s in self.socials],
        }


def normalize_query(name: str) -> str:
    """Normalize the query string."""
    return re.sub(r"\s+", " ", name.strip())


# OAuth token caching (valid ~1 hour)
_cached_token: Optional[Dict[str, any]] = None
_token_lock = asyncio.Lock()

async def get_ipau_access_token() -> str:
    """Get IP Australia OAuth access token with caching."""
    global _cached_token

    env = os.getenv("IPAU_ENV", "test")

    token_url = (
        os.getenv("IPAU_TOKEN_URL_PROD")
        if env == "production"
        else os.getenv("IPAU_TOKEN_URL_TEST")
    )

    client_id = os.getenv("IPAU_CLIENT_ID")
    client_secret = os.getenv("IPAU_CLIENT_SECRET")

    if not token_url:
        raise ValueError("Missing IPAU_TOKEN_URL_TEST/PROD in environment")
    if not client_id:
        raise ValueError("Missing IPAU_CLIENT_ID in environment")
    if not client_secret:
        raise ValueError("Missing IPAU_CLIENT_SECRET in environment")

    # If we have a valid cached token, reuse it
    if _cached_token and time.time() * 1000 < _cached_token["expiresAt"]:
        return _cached_token["token"]

    # Request new token
    async with _token_lock:
        return await request_new_token()

async def request_new_token() -> str:
    async with httpx.AsyncClient() as client:
        response = await client.post(
            token_url,
            data={
                "grant_type": "client_credentials",
                "client_id": client_id,
                "client_secret": client_secret,
            },
            headers={
                "Content-Type": "application/x-www-form-urlencoded",
                "Accept": "application/json",
            },
        )

        if not response.is_success:
            raise Exception(
                f"Token request failed {response.status_code}: {response.text[:250]}"
            )

        data = response.json()
        access_token = data.get("access_token")
        expires_in = int(data.get("expires_in", 3600))

        if not access_token:
            raise Exception(
                f"Token response missing access_token: {response.text[:250]}"
            )

        # Cache it with a small safety buffer (60s)
        _cached_token = {
            "token": access_token,
            "expiresAt": int(time.time() * 1000) + (expires_in - 60) * 1000,
        }

        return access_token

def extract_nice_classes(tm: Dict) -> List[str]:
    """Extract Nice class numbers from trademark data."""
    candidates = []

    # Direct arrays
    if isinstance(tm.get("classes"), list):
        candidates.extend(tm["classes"])
    if isinstance(tm.get("niceClasses"), list):
        candidates.extend(tm["niceClasses"])
    if isinstance(tm.get("classifications"), list):
        candidates.extend(tm["classifications"])

    # Nested common shapes
    goods_and_services = tm.get("goodsAndServices")
    if isinstance(goods_and_services, dict):
        if isinstance(goods_and_services.get("classes"), list):
            candidates.extend(goods_and_services["classes"])
        if isinstance(goods_and_services.get("niceClasses"), list):
            candidates.extend(goods_and_services["niceClasses"])

    # Sometimes it's goodsAndServices items, each with a class number
    if isinstance(goods_and_services, list):
        candidates.extend(goods_and_services)

    # Sometimes "classifications" is object with list inside
    if isinstance(tm.get("classificationList"), list):
        candidates.extend(tm["classificationList"])

    # Convert to class numbers
    numbers = []
    for c in candidates:
        if isinstance(c, dict):
            num = (
                c.get("number")
                or c.get("classNumber")
                or c.get("niceClass")
                or c.get("class")
                or c.get("classId")
            )
        else:
            num = c

        if num is not None:
            num_str = str(num).strip()
            if num_str:
                numbers.append(num_str)

    # Deduplicate + sort numerically when possible
    unique = sorted(set(numbers), key=lambda x: int(x) if x.isdigit() else 999)

    return unique


async def fetch_trademark_details(
    id: str, token: str, base_url: str
) -> Dict:
    """Fetch detailed trademark information."""
    url = f"{base_url.rstrip('/')}/trade-mark/{id}"

    async with httpx.AsyncClient() as client:
        response = await client.get(
            url,
            headers={
                "Authorization": f"Bearer {token}",
                "Accept": "application/json",
            },
        )

        if not response.is_success:
            return {
                "id": id,
                "error": f"Detail fetch failed {response.status_code}: {response.text[:120]}",
            }

        try:
            return {"id": id, "data": response.json()}
        except Exception:
            return {
                "id": id,
                "error": f"Detail fetch returned non-JSON: {response.text[:120]}",
            }


async def check_ip_australia_trademark(name: str) -> CheckResult:
    """Check IP Australia trademark availability."""
    q = normalize_query(name)

    env = os.getenv("IPAU_ENV", "test")
    base_url = (
        os.getenv("IPAU_TM_BASE_URL_PROD")
        if env == "production"
        else os.getenv("IPAU_TM_BASE_URL_TEST")
    )

    if not base_url:
        return CheckResult(
            label="Trademark (IP Australia)",
            status="unknown",
            summary="We couldn't run the trademark check right now.",
            why_this_matters="Trademark results help you avoid picking a name that could conflict with an existing brand.",
            details="Missing IPAU_TM_BASE_URL_TEST/PROD in environment",
        )

    url = f"{base_url.rstrip('/')}/search/quick"

    # Matches ApiQuickSearchRequest from the docs
    payload = {
        "query": q,
        "sort": {"field": "NUMBER", "direction": "ASCENDING"},
        "filters": {
            "quickSearchType": ["WORD"],
            "status": ["REGISTERED"],
        },
    }

    try:
        token = await get_ipau_access_token()

        # 1) Quick search
        async with httpx.AsyncClient() as client:
            response = await client.post(
                url,
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                },
                json=payload,
            )

            if not response.is_success:
                return CheckResult(
                    label="Trademark (IP Australia)",
                    status="unknown",
                    summary="We couldn't complete the trademark check right now.",
                    why_this_matters="Trademark results help you avoid choosing a name that could conflict with an existing brand.",
                    details=f"IP Australia error {response.status_code}: {response.text[:250]}",
                )

            data = response.json()

        # 2) Parse quick search result
        count = int(data.get("count", 0))
        trademark_ids = data.get("trademarkIds", [])

        if not isinstance(trademark_ids, list):
            trademark_ids = []

        if count == 0 or len(trademark_ids) == 0:
            return CheckResult(
                label="Trademark (IP Australia)",
                status="available",
                summary="No registered word trademarks were found in Australia.",
                why_this_matters="This lowers risk, but it's not a guarantee. Similar marks, pending applications, or other rights may still exist.",
                details="0 registered results returned by IP Australia quick search.",
            )

        # 3) Fetch details for top IDs (so we can show names + classes)
        top_ids = trademark_ids[:5]
        detail_results = await asyncio.gather(
            *[fetch_trademark_details(id, token, base_url) for id in top_ids]
        )

        # 4) Normalize detailed records -> structured matches
        q_lower = q.lower()

        parsed = []
        for d in detail_results:
            if "error" in d:
                parsed.append(
                    TrademarkMatch(
                        id=d["id"],
                        words="",
                        status="",
                        classes=[],
                        class_labels=[],
                    )
                )
                continue

            tm = d["data"]

            # words/name fields vary
            words = (
                tm.get("words")
                or tm.get("tradeMarkWords")
                or tm.get("markText")
                or tm.get("text")
                or tm.get("name")
                or tm.get("tradeMarkName")
                or ""
            )

            # status fields vary
            status = (
                tm.get("statusGroup")
                or tm.get("statusCode")
                or tm.get("statusDetail")
                or tm.get("status")
                or tm.get("tradeMarkStatus")
                or tm.get("state")
                or tm.get("ipRightStatus")
                or ""
            )

            # classes/categories can be nested; try common shapes
            classes_set = set()

            # IP Australia detail shape: goodsAndServices is an array of { class: "35", descriptionText: [...] }
            goods_and_services = tm.get("goodsAndServices")
            if isinstance(goods_and_services, list):
                for gs in goods_and_services:
                    cls = gs.get("class")
                    if cls:
                        classes_set.add(str(cls))

            classes = sorted(classes_set, key=lambda x: int(x) if x.isdigit() else 999)
            class_labels = [label_nice_class(c) for c in classes]

            parsed.append(
                TrademarkMatch(
                    id=d["id"],
                    words=str(words),
                    status=str(status),
                    classes=classes,
                    class_labels=class_labels,
                )
            )

        # Split into exact vs similar
        exact_matches = [
            TrademarkMatch(
                id=p.id,
                words=p.words,
                status=p.status,
                classes=p.classes,
                class_labels=p.class_labels,
            )
            for p in parsed
            if p.words and p.words.lower() == q_lower
        ]

        similar_matches = [
            TrademarkMatch(
                id=p.id,
                words=p.words,
                status=p.status,
                classes=p.classes,
                class_labels=p.class_labels,
            )
            for p in parsed
            if p.words and p.words.lower() != q_lower
        ]

        # If we couldn't parse any details, fall back to IDs only
        any_parsed = len(exact_matches) + len(similar_matches) > 0

        if not any_parsed:
            return CheckResult(
                label="Trademark (IP Australia)",
                status="similar",
                summary="Registered trademarks exist with this name or something similar.",
                why_this_matters="Even if names aren't identical, similar marks in related categories can increase legal and branding risk.",
                details=f"Found {count} registered match(es). IDs: {', '.join(trademark_ids[:5])}",
                similar_matches=[
                    TrademarkMatch(id=id) for id in trademark_ids[:5]
                ],
            )

        is_taken = len(exact_matches) > 0

        return CheckResult(
            label="Trademark (IP Australia)",
            status="taken" if is_taken else "similar",
            summary="An identical registered trademark exists in Australia."
            if is_taken
            else "Similar registered trademarks exist in Australia.",
            why_this_matters="Using the same name can create a high risk of trademark conflict, especially in related industries."
            if is_taken
            else "Even if names aren't identical, similar marks in related categories can still cause legal and branding risk.",
            details=f"Found {count} registered match(es). Showing the top {len(top_ids)}.",
            exact_matches=exact_matches,
            similar_matches=similar_matches,
        )

    except Exception as err:
        return CheckResult(
            label="Trademark (IP Australia)",
            status="unknown",
            summary="We couldn't run the trademark check right now.",
            why_this_matters="Trademark results help you avoid choosing a name that could conflict with an existing brand.",
            details=f"Request failed: {str(err)}",
        )


async def get_demo_results(name: str) -> AggregatedResults:
    """Get aggregated brand checking results (MVP with demo data)."""
    normalized = name.lower().strip()
    compact = normalized.replace(" ", "")

    # Demo logic (keep for now)
    is_probably_taken = "koala" in normalized or "australia" in normalized
    has_similar = "brew" in normalized or "coffee" in normalized

    # Real trademark check
    trademark = await check_ip_australia_trademark(name)

    return AggregatedResults(
        business_name=CheckResult(
            label="ASIC business name (AU)",
            status="taken" if is_probably_taken else "available",
            details="A similar or identical business name appears to exist (demo)."
            if is_probably_taken
            else "No exact match found (demo).",
        ),
        trademark=trademark,
        domains=[
            CheckResult(
                label=f"{compact}.com",
                status="taken" if is_probably_taken else "available",
                details="Likely registered already (demo)."
                if is_probably_taken
                else "Appears free (demo).",
            ),
            CheckResult(
                label=f"{compact}.com.au",
                status="similar" if has_similar else "available",
                details="Similar domains may exist (demo)."
                if has_similar
                else "Appears free (demo).",
            ),
        ],
        socials=[
            CheckResult(
                label=f"@{compact} (Instagram)",
                status="taken" if is_probably_taken else "available",
                details="Handle looks popular (demo)."
                if is_probably_taken
                else "Appears free (demo).",
            ),
            CheckResult(
                label=f"@{compact} (TikTok)",
                status="available",
                details="Appears free (demo).",
            ),
        ],
    )
